File: python/Main/warn_logic.py
from bot_warnings import (
    add_warning,
    get_user_warnings,
    remove_warnings,
    load_warnings,
    save_warnings,
)
import discord # type: ignore
from datetime import datetime, timedelta
from quarantine import get_log_channel
from bot_setup import bot


# Helper function for warn logic
async def warn_user_logic(user, moderator, reason, guild, response_func):
    """Shared logic for warning users (used by both prefix and slash commands)"""
    # Prevent warning bots or yourself
    if user.bot:
        await response_func("❌ You cannot warn bots!")
        return

    if user.id == moderator.id:
        await response_func("❌ You cannot warn yourself!")
        return

    # Add warning to database
    warning_count = add_warning(user.id, moderator.id, reason, guild.id)

    # Create warning embed
    embed = discord.Embed(
        title="⚠️ User Warned", color=discord.Color.orange(), timestamp=datetime.now()
    )
    embed.add_field(name="User", value=f"{user.mention} ({user})", inline=True)
    embed.add_field(name="Moderator", value=f"{moderator.mention}", inline=True)
    embed.add_field(name="Warning Count", value=f"{warning_count}", inline=True)
    embed.add_field(name="Reason", value=reason, inline=False)

    # Log the warning action
    log_channel_id = get_log_channel(guild.id)
    if log_channel_id:
        log_channel = guild.get_channel(log_channel_id)
        if log_channel:
            log_embed = discord.Embed(
                title="⚠️ Warning Issued",
                color=discord.Color.orange(),
                timestamp=datetime.now(),
            )
            log_embed.add_field(
                name="User", value=f"{user.mention} ({user})", inline=True
            )
            log_embed.add_field(
                name="Moderator", value=f"{moderator.mention}", inline=True
            )
            log_embed.add_field(
                name="Warning Count", value=f"{warning_count}", inline=True
            )
            log_embed.add_field(name="Reason", value=reason, inline=False)
            await log_channel.send(embed=log_embed)

    # Auto-timeout and action logic
    action_taken = "Warning issued"
    timeout_duration = None
    kick_user = False

    if warning_count == 2:
        timeout_duration = timedelta(minutes=30)
        action_taken = "⏰ 30 minute timeout applied (2nd warning)"
    elif warning_count == 3:
        timeout_duration = timedelta(hours=3)
        action_taken = "⏰ 3 hour timeout applied (3rd warning)"
    elif warning_count == 4:
        timeout_duration = timedelta(days=7)
        action_taken = "⏰ 7 day timeout applied (4th warning)"
    elif warning_count >= 5:
        kick_user = True
        action_taken = "🚫 User kicked (5th warning)"

    # Then apply punishment
    if timeout_duration:
        try:
            await user.timeout(
                timeout_duration, reason=f"Auto-timeout: {warning_count} warnings"
            )
            embed.add_field(name="Action Taken", value=action_taken, inline=False)
            embed.color = discord.Color.red()

            # Log the timeout action
            if log_channel_id:
                log_channel = guild.get_channel(log_channel_id)
                if log_channel:
                    timeout_embed = discord.Embed(
                        title="⏰ Auto-Timeout Applied",
                        color=discord.Color.red(),
                        timestamp=datetime.now(),
                    )
                    timeout_embed.add_field(
                        name="User", value=f"{user.mention} ({user})", inline=True
                    )
                    timeout_embed.add_field(
                        name="Moderator", value="Auto-moderation", inline=True
                    )
                    timeout_embed.add_field(
                        name="Duration", value=str(timeout_duration), inline=True
                    )
                    timeout_embed.add_field(
                        name="Reason",
                        value=f"Auto-timeout: {warning_count} warnings",
                        inline=False,
                    )
                    await log_channel.send(embed=timeout_embed)

        except discord.Forbidden:
            embed.add_field(
                name="⚠️ Error",
                value="Could not apply timeout - insufficient permissions",
                inline=False,
            )
        except Exception as e:
            embed.add_field(
                name="⚠️ Error", value=f"Failed to apply timeout: {str(e)}", inline=False
            )
    elif kick_user:
        try:
            await user.kick(reason=f"Auto-kick: {warning_count} warnings")
            embed.add_field(name="Action Taken", value=action_taken, inline=False)
            embed.color = discord.Color.red()

            # Log the kick action
            if log_channel_id:
                log_channel = guild.get_channel(log_channel_id)
                if log_channel:
                    kick_embed = discord.Embed(
                        title="🚫 Auto-Kick Applied",
                        color=discord.Color.red(),
                        timestamp=datetime.now(),
                    )
                    kick_embed.add_field(
                        name="User", value=f"{user.mention} ({user})", inline=True
                    )
                    kick_embed.add_field(
                        name="Moderator", value="Auto-moderation", inline=True
                    )
                    kick_embed.add_field(
                        name="Reason",
                        value=f"Auto-kick: {warning_count} warnings",
                        inline=False,
                    )
                    await log_channel.send(embed=kick_embed)

        except discord.Forbidden:
            embed.add_field(
                name="⚠️ Error",
                value="Could not kick user - insufficient permissions",
                inline=False,
            )
        except Exception as e:
            embed.add_field(
                name="⚠️ Error", value=f"Failed to kick user: {str(e)}", inline=False
            )

    await response_func(embed=embed)


# Helper function for warnings check logic
async def check_warnings_logic(user, moderator, response_func, ephemeral=False):
    """Shared logic for checking warnings (used by both prefix and slash commands)"""
    warnings = get_user_warnings(user.id)

    embed = discord.Embed(
        title=f"📋 Warnings for {user.display_name}",
        color=discord.Color.blue(),
        timestamp=datetime.now(),
    )
    embed.set_thumbnail(url=user.display_avatar.url)

    if not warnings:
        embed.description = "This user has no warnings."
        embed.color = discord.Color.green()
    else:
        embed.description = f"Total warnings: **{len(warnings)}**"

        # Show last 5 warnings
        recent_warnings = warnings[-5:]
        for i, warning in enumerate(recent_warnings, 1):
            moderator_user = bot.get_user(warning["moderator_id"])
            mod_name = (
                moderator_user.display_name if moderator_user else "Unknown Moderator"
            )

            # Parse timestamp
            warning_time = datetime.fromisoformat(warning["timestamp"])
            time_str = warning_time.strftime("%Y-%m-%d %H:%M UTC")

            embed.add_field(
                name=f"Warning {len(warnings) - len(recent_warnings) + i}",
                value=f"**Reason:** {warning['reason']}\n**Moderator:** {mod_name}\n**Date:** {time_str}",
                inline=False,
            )

        if len(warnings) > 5:
            embed.set_footer(text=f"Showing latest 5 of {len(warnings)} warnings")

    if ephemeral:
        await response_func(embed=embed, ephemeral=True)
    else:
        await response_func(embed=embed)


# Helper function for remove warnings logic
async def remove_warnings_logic(
    user, moderator, guild, amount, response_func, ephemeral=False
):
    """Shared logic for removing specific amount of warnings"""
    removed, remaining = remove_warnings(user.id, amount)

    if removed == 0:
        error_msg = f"❌ {user.display_name} has no warnings to remove!"
        if ephemeral:
            await response_func(error_msg, ephemeral=True)
        else:
            await response_func(error_msg)
        return

    embed = discord.Embed(
        title="🗑️ Warnings Removed",
        description=f"Removed **{removed}** warning(s) from {user.mention}",
        color=discord.Color.yellow(),
        timestamp=datetime.now(),
    )
    embed.add_field(name="Moderator", value=moderator.mention, inline=True)
    embed.add_field(name="Remaining Warnings", value=f"{remaining}", inline=True)

    # Log the warning removal action
    log_channel_id = get_log_channel(guild.id)
    if log_channel_id:
        log_channel = guild.get_channel(log_channel_id)
        if log_channel:
            log_embed = discord.Embed(
                title="🗑️ Warnings Removed",
                color=discord.Color.yellow(),
                timestamp=datetime.now(),
            )
            log_embed.add_field(
                name="User", value=f"{user.mention} ({user})", inline=True
            )
            log_embed.add_field(name="Moderator", value=moderator.mention, inline=True)
            log_embed.add_field(name="Removed", value=f"{removed}", inline=True)
            log_embed.add_field(name="Remaining", value=f"{remaining}", inline=True)
            await log_channel.send(embed=log_embed)

    if ephemeral:
        await response_func(embed=embed, ephemeral=True)
    else:
        await response_func(embed=embed)


async def clear_warnings_logic(user, moderator, guild, response_func, ephemeral=False):
    """Shared logic for clearing warnings (used by both prefix and slash commands)"""
    warnings_data = load_warnings()
    user_id_str = str(user.id)

    if user_id_str not in warnings_data or not warnings_data[user_id_str]:
        error_msg = f"❌ {user.display_name} has no warnings to clear!"
        if ephemeral:
            await response_func(error_msg, ephemeral=True)
        else:
            await response_func(error_msg)
        return

    old_count = len(warnings_data[user_id_str])
    warnings_data[user_id_str] = []
    save_warnings(warnings_data)

    embed = discord.Embed(
        title="🗑️ Warnings Cleared",
        description=f"Cleared **{old_count}** warning(s) for {user.mention}",
        color=discord.Color.green(),
        timestamp=datetime.now(),
    )
    embed.add_field(name="Moderator", value=moderator.mention, inline=True)

    # Log the warning clearing action
    log_channel_id = get_log_channel(guild.id)
    if log_channel_id:
        log_channel = guild.get_channel(log_channel_id)
        if log_channel:
            log_embed = discord.Embed(
                title="🗑️ Warnings Cleared",
                color=discord.Color.green(),
                timestamp=datetime.now(),
            )
            log_embed.add_field(
                name="User", value=f"{user.mention} ({user})", inline=True
            )
            log_embed.add_field(name="Moderator", value=moderator.mention, inline=True)
            log_embed.add_field(name="Cleared", value=f"{old_count}", inline=True)
            await log_channel.send(embed=log_embed)

    if ephemeral:
        await response_func(embed=embed, ephemeral=True)
    else:
        await response_func(embed=embed)


# Helper function for remove timeout logic
async def remove_timeout_logic(user, moderator, response_func, ephemeral=False):
    """Shared logic for removing timeouts (used by both prefix and slash commands)"""
    # Check if user is actually timed out
    if user.timed_out_until is None:
        error_msg = f"❌ {user.display_name} is not currently timed out!"
        if ephemeral:
            await response_func(error_msg, ephemeral=True)
        else:
            await response_func(error_msg)
        return

    try:
        # Remove timeout by setting it to None
        await user.timeout(None, reason=f"Timeout removed by {moderator}")

        # Send confirmation to moderator
        embed = discord.Embed(
            title="⏰ Timeout Removed",
            description=f"Timeout removed for {user.mention}",
            color=discord.Color.green(),
            timestamp=datetime.now(),
        )
        embed.add_field(name="Moderator", value=moderator.mention, inline=True)
        embed.add_field(name="User", value=f"{user.mention} ({user})", inline=True)

        # Log the timeout removal action
        log_channel_id = get_log_channel(moderator.guild.id)
        if log_channel_id:
            log_channel = moderator.guild.get_channel(log_channel_id)
            if log_channel:
                log_embed = discord.Embed(
                    title="⏰ Timeout Removed",
                    color=discord.Color.green(),
                    timestamp=datetime.now(),
                )
                log_embed.add_field(
                    name="User", value=f"{user.mention} ({user})", inline=True
                )
                log_embed.add_field(
                    name="Moderator", value=moderator.mention, inline=True
                )
                log_embed.add_field(
                    name="Reason", value=f"Timeout removed by {moderator}", inline=False
                )
                await log_channel.send(embed=log_embed)

        if ephemeral:
            await response_func(embed=embed, ephemeral=True)
        else:
            await response_func(embed=embed)

    except discord.Forbidden:
        error_msg = "❌ Could not remove timeout - insufficient permissions!"
        if ephemeral:
            await response_func(error_msg, ephemeral=True)
        else:
            await response_func(error_msg)
    except Exception as e:
        error_msg = f"❌ Failed to remove timeout: {str(e)}"
        if ephemeral:
            await response_func(error_msg, ephemeral=True)
        else:
            await response_func(error_msg)

File: nodejs/src/modules/warnLogic.js
import { EmbedBuilder } from 'discord.js';
import { addWarning, getUserWarnings, removeWarnings, loadWarnings, saveWarnings } from './warnings.js';
import { getLogChannel } from './quarantine.js';
import { client } from '../index.js';

// Timeout durations in milliseconds
const TIMEOUT_30_MIN = 30 * 60 * 1000;
const TIMEOUT_3_HOURS = 3 * 60 * 60 * 1000;
const TIMEOUT_7_DAYS = 7 * 24 * 60 * 60 * 1000;

/**
 * Shared logic for warning users
 */
export async function warnUserLogic(user, moderator, reason, guild, responseFunc) {
    // Prevent warning bots or yourself
    if (user.bot) {
        await responseFunc({ content: '❌ You cannot warn bots!' });
        return;
    }

    if (user.id === moderator.id) {
        await responseFunc({ content: '❌ You cannot warn yourself!' });
        return;
    }

    // Add warning to database
    const warningCount = addWarning(user.id, moderator.id, reason, guild.id);

    // Create warning embed
    const embed = new EmbedBuilder()
        .setTitle('⚠️ User Warned')
        .setColor(0xFFA500)
        .setTimestamp()
        .addFields(
            { name: 'User', value: `${user} (${user.tag})`, inline: true },
            { name: 'Moderator', value: `${moderator}`, inline: true },
            { name: 'Warning Count', value: `${warningCount}`, inline: true },
            { name: 'Reason', value: reason, inline: false }
        );

    // Log the warning action
    const logChannelId = getLogChannel(guild.id);
    if (logChannelId) {
        const logChannel = guild.channels.cache.get(logChannelId);
        if (logChannel) {
            const logEmbed = new EmbedBuilder()
                .setTitle('⚠️ Warning Issued')
                .setColor(0xFFA500)
                .setTimestamp()
                .addFields(
                    { name: 'User', value: `${user} (${user.tag})`, inline: true },
                    { name: 'Moderator', value: `${moderator}`, inline: true },
                    { name: 'Warning Count', value: `${warningCount}`, inline: true },
                    { name: 'Reason', value: reason, inline: false }
                );
            await logChannel.send({ embeds: [logEmbed] });
        }
    }

    // Auto-timeout and action logic
    let actionTaken = 'Warning issued';
    let timeoutDuration = null;
    let kickUser = false;

    if (warningCount === 2) {
        timeoutDuration = TIMEOUT_30_MIN;
        actionTaken = '⏰ 30 minute timeout applied (2nd warning)';
    } else if (warningCount === 3) {
        timeoutDuration = TIMEOUT_3_HOURS;
        actionTaken = '⏰ 3 hour timeout applied (3rd warning)';
    } else if (warningCount === 4) {
        timeoutDuration = TIMEOUT_7_DAYS;
        actionTaken = '⏰ 7 day timeout applied (4th warning)';
    } else if (warningCount >= 5) {
        kickUser = true;
        actionTaken = '🚫 User kicked (5th warning)';
    }

    // Apply punishment
    const member = guild.members.cache.get(user.id);
    
    if (timeoutDuration && member) {
        try {
            await member.timeout(timeoutDuration, `Auto-timeout: ${warningCount} warnings`);
            embed.addFields({ name: 'Action Taken', value: actionTaken, inline: false });
            embed.setColor(0xFF0000);

            // Log the timeout action
            if (logChannelId) {
                const logChannel = guild.channels.cache.get(logChannelId);
                if (logChannel) {
                    const durationStr = warningCount === 2 ? '30 minutes' : 
                                       warningCount === 3 ? '3 hours' : '7 days';
                    const timeoutEmbed = new EmbedBuilder()
                        .setTitle('⏰ Auto-Timeout Applied')
                        .setColor(0xFF0000)
                        .setTimestamp()
                        .addFields(
                            { name: 'User', value: `${user} (${user.tag})`, inline: true },
                            { name: 'Moderator', value: 'Auto-moderation', inline: true },
                            { name: 'Duration', value: durationStr, inline: true },
                            { name: 'Reason', value: `Auto-timeout: ${warningCount} warnings`, inline: false }
                        );
                    await logChannel.send({ embeds: [timeoutEmbed] });
                }
            }
        } catch (error) {
            embed.addFields({ name: '⚠️ Error', value: 'Could not apply timeout - insufficient permissions', inline: false });
        }
    } else if (kickUser && member) {
        try {
            await member.kick(`Auto-kick: ${warningCount} warnings`);
            embed.addFields({ name: 'Action Taken', value: actionTaken, inline: false });
            embed.setColor(0xFF0000);

            // Log the kick action
            if (logChannelId) {
                const logChannel = guild.channels.cache.get(logChannelId);
                if (logChannel) {
                    const kickEmbed = new EmbedBuilder()
                        .setTitle('🚫 Auto-Kick Applied')
                        .setColor(0xFF0000)
                        .setTimestamp()
                        .addFields(
                            { name: 'User', value: `${user} (${user.tag})`, inline: true },
                            { name: 'Moderator', value: 'Auto-moderation', inline: true },
                            { name: 'Reason', value: `Auto-kick: ${warningCount} warnings`, inline: false }
                        );
                    await logChannel.send({ embeds: [kickEmbed] });
                }
            }
        } catch (error) {
            embed.addFields({ name: '⚠️ Error', value: 'Could not kick user - insufficient permissions', inline: false });
        }
    }

    await responseFunc({ embeds: [embed] });
}

/**
 * Shared logic for checking warnings
 */
export async function checkWarningsLogic(user, moderator, responseFunc, ephemeral = false) {
    const warnings = getUserWarnings(user.id);

    const embed = new EmbedBuilder()
        .setTitle(`📋 Warnings for ${user.displayName || user.username}`)
        .setColor(0x0000FF)
        .setTimestamp()
        .setThumbnail(user.displayAvatarURL());

    if (!warnings.length) {
        embed.setDescription('This user has no warnings.');
        embed.setColor(0x00FF00);
    } else {
        embed.setDescription(`Total warnings: **${warnings.length}**`);

        // Show last 5 warnings
        const recentWarnings = warnings.slice(-5);
        for (let i = 0; i < recentWarnings.length; i++) {
            const warning = recentWarnings[i];
            let modName = 'Unknown Moderator';
            try {
                const moderatorUser = await client.users.fetch(warning.moderator_id);
                modName = moderatorUser.displayName || moderatorUser.username;
            } catch (e) {
                // Ignore fetch errors
            }

            const warningTime = new Date(warning.timestamp);
            const timeStr = warningTime.toISOString().replace('T', ' ').substring(0, 19) + ' UTC';

            embed.addFields({
                name: `Warning ${warnings.length - recentWarnings.length + i + 1}`,
                value: `**Reason:** ${warning.reason}\n**Moderator:** ${modName}\n**Date:** ${timeStr}`,
                inline: false
            });
        }

        if (warnings.length > 5) {
            embed.setFooter({ text: `Showing latest 5 of ${warnings.length} warnings` });
        }
    }

    await responseFunc({ embeds: [embed], ephemeral });
}

/**
 * Shared logic for removing specific amount of warnings
 */
export async function removeWarningsLogic(user, moderator, guild, amount, responseFunc, ephemeral = false) {
    const { removed, remaining } = removeWarnings(user.id, amount);

    if (removed === 0) {
        const errorMsg = `❌ ${user.displayName || user.username} has no warnings to remove!`;
        await responseFunc({ content: errorMsg, ephemeral });
        return;
    }

    const embed = new EmbedBuilder()
        .setTitle('🗑️ Warnings Removed')
        .setDescription(`Removed **${removed}** warning(s) from ${user}`)
        .setColor(0xFFFF00)
        .setTimestamp()
        .addFields(
            { name: 'Moderator', value: `${moderator}`, inline: true },
            { name: 'Remaining Warnings', value: `${remaining}`, inline: true }
        );

    // Log the warning removal action
    const logChannelId = getLogChannel(guild.id);
    if (logChannelId) {
        const logChannel = guild.channels.cache.get(logChannelId);
        if (logChannel) {
            const logEmbed = new EmbedBuilder()
                .setTitle('🗑️ Warnings Removed')
                .setColor(0xFFFF00)
                .setTimestamp()
                .addFields(
                    { name: 'User', value: `${user} (${user.tag})`, inline: true },
                    { name: 'Moderator', value: `${moderator}`, inline: true },
                    { name: 'Removed', value: `${removed}`, inline: true },
                    { name: 'Remaining', value: `${remaining}`, inline: true }
                );
            await logChannel.send({ embeds: [logEmbed] });
        }
    }

    await responseFunc({ embeds: [embed], ephemeral });
}

/**
 * Shared logic for clearing warnings
 */
export async function clearWarningsLogic(user, moderator, guild, responseFunc, ephemeral = false) {
    const warningsData = loadWarnings();
    const userIdStr = String(user.id);

    if (!warningsData[userIdStr] || !warningsData[userIdStr].length) {
        const errorMsg = `❌ ${user.displayName || user.username} has no warnings to clear!`;
        await responseFunc({ content: errorMsg, ephemeral });
        return;
    }

    const oldCount = warningsData[userIdStr].length;
    warningsData[userIdStr] = [];
    saveWarnings(warningsData);

    const embed = new EmbedBuilder()
        .setTitle('🗑️ Warnings Cleared')
        .setDescription(`Cleared **${oldCount}** warning(s) for ${user}`)
        .setColor(0x00FF00)
        .setTimestamp()
        .addFields({ name: 'Moderator', value: `${moderator}`, inline: true });

    // Log the warning clearing action
    const logChannelId = getLogChannel(guild.id);
    if (logChannelId) {
        const logChannel = guild.channels.cache.get(logChannelId);
        if (logChannel) {
            const logEmbed = new EmbedBuilder()
                .setTitle('🗑️ Warnings Cleared')
                .setColor(0x00FF00)
                .setTimestamp()
                .addFields(
                    { name: 'User', value: `${user} (${user.tag})`, inline: true },
                    { name: 'Moderator', value: `${moderator}`, inline: true },
                    { name: 'Cleared', value: `${oldCount}`, inline: true }
                );
            await logChannel.send({ embeds: [logEmbed] });
        }
    }

    await responseFunc({ embeds: [embed], ephemeral });
}

/**
 * Shared logic for removing timeouts
 */
export async function removeTimeoutLogic(user, moderator, responseFunc, ephemeral = false) {
    const member = moderator.guild.members.cache.get(user.id);
    
    if (!member || !member.communicationDisabledUntil) {
        const errorMsg = `❌ ${user.displayName || user.username} is not currently timed out!`;
        await responseFunc({ content: errorMsg, ephemeral });
        return;
    }

    try {
        await member.timeout(null, `Timeout removed by ${moderator.tag}`);

        const embed = new EmbedBuilder()
            .setTitle('⏰ Timeout Removed')
            .setDescription(`Timeout removed for ${user}`)
            .setColor(0x00FF00)
            .setTimestamp()
            .addFields(
                { name: 'Moderator', value: `${moderator}`, inline: true },
                { name: 'User', value: `${user} (${user.tag})`, inline: true }
            );

        // Log the timeout removal action
        const logChannelId = getLogChannel(moderator.guild.id);
        if (logChannelId) {
            const logChannel = moderator.guild.channels.cache.get(logChannelId);
            if (logChannel) {
                const logEmbed = new EmbedBuilder()
                    .setTitle('⏰ Timeout Removed')
                    .setColor(0x00FF00)
                    .setTimestamp()
                    .addFields(
                        { name: 'User', value: `${user} (${user.tag})`, inline: true },
                        { name: 'Moderator', value: `${moderator}`, inline: true },
                        { name: 'Reason', value: `Timeout removed by ${moderator.tag}`, inline: false }
                    );
                await logChannel.send({ embeds: [logEmbed] });
            }
        }

        await responseFunc({ embeds: [embed], ephemeral });

    } catch (error) {
        if (error.code === 50013) {
            await responseFunc({ content: "❌ Could not remove timeout - insufficient permissions!", ephemeral });
        } else {
            await responseFunc({ content: `❌ Failed to remove timeout: ${error.message}`, ephemeral });
        }
    }
}
